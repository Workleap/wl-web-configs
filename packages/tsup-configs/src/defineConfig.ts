import { defineConfig as _defineConfig, type Options } from "tsup";

export type DefineConfigOptions = Parameters<typeof _defineConfig>[0];

export function defineConfig(baseOptions: Options, userOptions?: DefineConfigOptions): ReturnType<typeof _defineConfig> {
    if (!userOptions || (Array.isArray(userOptions) && userOptions.length === 0)) {
        return _defineConfig(baseOptions);
    }

    if (typeof userOptions === "function") {
        return _defineConfig((...args) => ({
            ...baseOptions,
            ...userOptions(...args)
        }));
    }

    if (Array.isArray(userOptions)) {
        return _defineConfig(
            userOptions.map(configItem => {
                return {
                    ...baseOptions,
                    ...configItem
                };
            })
        );
    }

    return _defineConfig({
        ...baseOptions,
        ...userOptions
    });
}
