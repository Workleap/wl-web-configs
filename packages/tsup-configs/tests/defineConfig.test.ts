import type { Options } from "tsup";
import { defineConfig } from "../src/defineConfig.ts";

test("when no options are provided, return the default options", () => {
    const baseOptions: Options = {
        dts: true
    };

    const result = defineConfig(baseOptions);

    expect(result).toEqual(baseOptions);
});

describe("single object of options", () => {
    test("when options are provided, options are merged with the default options", () => {
        const baseOptions: Options = {
            dts: true
        };

        const result = defineConfig(baseOptions, {
            clean: true
        });

        expect(result).toMatchSnapshot();
    });

    test("when a provided option match a default option, override the default option", () => {
        const baseOptions: Options = {
            dts: true,
            entry: ["./src"]
        };

        const result = defineConfig(baseOptions, {
            dts: false,
            entry: ["./toto"]
        });

        expect(result).toMatchSnapshot();
    });

    test("when a format option is provided, do not merge the provided array with the default format array", () => {
        const baseOptions: Options = {
            format: ["esm"]
        };

        const result = defineConfig(baseOptions, {
            format: ["cjs"]
        });

        expect(result).toMatchSnapshot();
    });
});

describe("array of options objects", () => {
    test("when an empty array is provided, returns the default options", () => {
        const baseOptions: Options = {
            dts: true
        };

        const result = defineConfig(baseOptions);

        expect(result).toEqual(baseOptions);
    });

    test("when options are provided, every object included in the array are merged with the default options", () => {
        const baseOptions: Options = {
            dts: true
        };

        const result = defineConfig(baseOptions, [
            { clean: true },
            { clean: false }
        ]);

        expect(result).toMatchSnapshot();
    });

    test("when a provided option match a default option, override the default option", () => {
        const baseOptions: Options = {
            dts: true,
            entry: ["./src"]
        };

        const result = defineConfig(baseOptions, [
            { dts: false, entry: ["./toto"] },
            { clean: false, entry: ["./toto"] }
        ]);

        expect(result).toMatchSnapshot();
    });

    test("when a format option is provided, do not merge the provided array with the default format array", () => {
        const baseOptions: Options = {
            format: ["esm"]
        };

        const result = defineConfig(baseOptions, [
            { format: ["cjs"] }
        ]);

        expect(result).toMatchSnapshot();
    });
});

describe("function", () => {
    test("when a function is provided, merge the options returned by the function with the default options", () => {
        const baseOptions: Options = {
            dts: true
        };

        const result = defineConfig(baseOptions, () => ({
            clean: true
        })) as () => unknown;

        expect(result()).toMatchSnapshot();
    });

    test("when a provided option match a default option, override the default option", () => {
        const baseOptions: Options = {
            dts: true,
            entry: ["./src"]
        };

        const result = defineConfig(baseOptions, () => ({
            dts: false,
            entry: ["./toto"]
        })) as () => unknown;

        expect(result()).toMatchSnapshot();
    });

    test("when a format option is provided, do not merge the provided array with the default format array", () => {
        const baseOptions: Options = {
            format: ["esm"]
        };

        const result = defineConfig(baseOptions, () => ({
            format: ["cjs"]
        })) as () => unknown;

        expect(result()).toMatchSnapshot();
    });
});
