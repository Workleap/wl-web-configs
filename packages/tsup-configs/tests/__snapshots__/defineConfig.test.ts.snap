// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array of options objects when a format option is provided, do not merge the provided array with the default format array 1`] = `
[
  {
    "format": [
      "cjs",
    ],
  },
]
`;

exports[`array of options objects when a provided option match a default option, override the default option 1`] = `
[
  {
    "dts": false,
    "entry": [
      "./toto",
    ],
  },
  {
    "clean": false,
    "dts": true,
    "entry": [
      "./toto",
    ],
  },
]
`;

exports[`array of options objects when options are provided, every object included in the array are merged with the default options 1`] = `
[
  {
    "clean": true,
    "dts": true,
  },
  {
    "clean": false,
    "dts": true,
  },
]
`;

exports[`function when a format option is provided, do not merge the provided array with the default format array 1`] = `
{
  "format": [
    "cjs",
  ],
}
`;

exports[`function when a function is provided, merge the options returned by the function with the default options 1`] = `
{
  "clean": true,
  "dts": true,
}
`;

exports[`function when a provided option match a default option, override the default option 1`] = `
{
  "dts": false,
  "entry": [
    "./toto",
  ],
}
`;

exports[`single object of options when a format option is provided, do not merge the provided array with the default format array 1`] = `
{
  "format": [
    "cjs",
  ],
}
`;

exports[`single object of options when a provided option match a default option, override the default option 1`] = `
{
  "dts": false,
  "entry": [
    "./toto",
  ],
}
`;

exports[`single object of options when options are provided, options are merged with the default options 1`] = `
{
  "clean": true,
  "dts": true,
}
`;
